juju_verify/verifiers/neutron_gateway.py

```diff
@@ -0,0 +1,162 @@
+        """Check that there are no non-redundant resources matching the resource type."""
+        shutdown_resource_list = self.get_shutdown_resource_list(action_name)
+        redundant_resource_list = self.get_online_resource_list(action_name)
+
+        shutdown_resource_set = set(r["id"] for r in shutdown_resource_list)
+        redundant_resource_set = set(r["id"] for r in redundant_resource_list)
+        non_redundant_list = shutdown_resource_set - redundant_resource_set
+        if non_redundant_list:
+            failure_string = self.action_name_failure_string_map[action_name]
+            reason = failure_string.format(", ".join(non_redundant_list))
+            result = Result(Severity.FAIL, reason)
+        else:
+            resource = self.action_name_result_map[action_name]
+            reason = "Redundancy check passed for: {}".format(resource)
+            result = Result(Severity.OK, reason)
+        return result
+
+    def warn_router_ha(self) -> Result:
+        """Warn that HA routers should be manually failed over."""
+        result = Result()
+        shutdown_resource_list = self.get_shutdown_resource_list("get-status-routers")
+
+        router_failover_err_list = []
+        for router in shutdown_resource_list:
+            if router["ha"]:
+                _id = router["id"]
+                entity_id = router["juju-entity-id"]
+                host = router["host"]
+                error_string = "{} (on {}, hostname: {})".format(_id, entity_id, host)
+                router_failover_err_list.append(error_string)
+
+        if router_failover_err_list:
+            error_string = ("It's recommended that you manually failover the following "
+                            "routers: {}")
+            reason = error_string.format(", ".join(router_failover_err_list))
+            result = Result(Severity.WARN, reason)
+
+        return result
+
+    def warn_lbaas_present(self) -> Result:
+        """Warn that LBaasV2 loadbalancers are present on the verified units."""
+        result = Result()
+        shutdown_lbaas_list = self.get_resource_list("get-status-lb")
+        units_with_lbaas = {unit["juju-entity-id"] for unit in shutdown_lbaas_list}
+        affected_lbaas_units = set(self.unit_ids) & units_with_lbaas
+
+        if affected_lbaas_units:
+            message = ('Following units have neutron LBaasV2 load-balancers that will be'
+                       ' lost on unit shutdown: {}')
+            reason = message.format(", ".join(affected_lbaas_units))
+            result = Result(Severity.WARN, reason)
+
+        return result
+
+    def verify_reboot(self) -> Result:
+        """Verify that it's safe to reboot selected neutron-gateway units."""
+        return self.verify_shutdown()
+
+    def verify_shutdown(self) -> Result:
+        """Verify that it's safe to shutdown selected neutron-gateway units."""
+        return aggregate_results(self.warn_router_ha(),
```

Comment:
Due to future `checks_executor` function #14, could you create `self.check_non_redundant_routers()` and `self.check_non_redundant_dhcp()`?

Reply:
huh...this is actually something that we didn't think of and I think it's problem with `checks_executor` design. Not having an option to execute checks with arguments is oversight imo.

Reply:
There is a way to allow this to work with parameters, but I think the checks will be defined much cleaner without any parameters.

Reply:
I agree about the cleanliness, and I also agree about the limitation. What if methods without parameters would call the generic methods with parameters? `self.check_non_redundance_routers` would call `self.check_non_redundant_resource("get-status-routers")`.

Alternatively, `checks_executor` could get lists of one (no params) or more (list[1:] would be the params) parameters.

Reply:
Yes that was my point, something like this.
```python
def check_non_redundance_routers(self):
    "Nice docstring about the check."
    return self.check_non_redundant_resource("get-status-routers")
```